packages/effect/src/internal/schedule.ts

```diff
@@ -1886,11 +1938,17 @@ const repeatOrElseEffectLoop = <A, E, R, R1, B, C, E2, R2>(
 ): Effect.Effect<B | C, E2, R | R1 | R2> => {
   return core.matchEffect(driver.next(value), {
     onFailure: () => core.orDie(driver.last),
-    onSuccess: (b) =>
-      core.matchEffect(self, {
-        onFailure: (error) => orElse(error, Option.some(b)),
-        onSuccess: (value) => repeatOrElseEffectLoop(self, driver, orElse, value)
+    onSuccess: (b) => {
+      const provideLastIterationInfo = effect.provideServiceEffect(
+        CurrentIterationMetadata,
+        ref.get(driver.iterationMeta)
+      )
+      const selfWithLastIterationInfo = provideLastIterationInfo(self)
+      return core.matchEffect(selfWithLastIterationInfo, {
+        onFailure: (error) => provideLastIterationInfo(orElse(error, Option.some(b))),
+        onSuccess: (value) => repeatOrElseEffectLoop(selfWithLastIterationInfo, driver, orElse, value)
```

Comment:
```suggestion
        onSuccess: (value) => repeatOrElseEffectLoop(self, driver, orElse, value)
```

Otherwise the effect definition could get infinitely large.